### PR TITLE
Job application review page form

### PR DESCRIPTION
**Pull request by cpjmcquillan**
Some things I wanted to get in for now. Some parts are placeholders until the tickets get picked up shortly.

The main thing is the idea that a job application has completed steps. It is very possible that we will want to know this information, and we definitely want to ensure all steps are complete before submitting applications. It will have the added benefit of cleaning up some of the "which steps are complete?" sidebar logic. The final details of how we present incomplete steps to the user needs to be ironed out. 

I think this approach is an improvement on the `Create a job listing` journey where we check that all forms are valid (it's a bit nasty imo)

### CODE REVIEW

File: app/controllers/jobseekers/job_applications_controller.rb [hunk 1]
```diff
@@ -1,19 +1,30 @@
...
     request_event.trigger(:vacancy_apply_clicked, vacancy_id: vacancy.id)
   end
 
   def create
     new_job_application = current_jobseeker.job_applications.find_or_initialize_by(vacancy: vacancy)
-    new_job_application.draft!
-    redirect_to jobseekers_job_application_build_path(new_job_application, :personal_details)
+
+    if new_job_application.persisted?
+      redirect_to jobseeker_root_path, notice: "You already applied for this job"
```
Comment (david-mears-dfe) [inline app/controllers/jobseekers/job_applications_controller.rb:12]: Where do these pieces of content come from? I believe this one should be 'You have already'.

File: app/models/job_application.rb [hunk 2]
```diff
@@ -1,4 +1,16 @@
 class JobApplication < ApplicationRecord
+  extend ArrayEnum
+
+  array_enum completed_steps: {
```
Comment (david-mears-dfe) [inline app/models/job_application.rb:4]: Why no 1/6?

File: app/controllers/jobseekers/job_applications_controller.rb (hunk 1 above)
Comment (cpjmcquillan) [reply app/controllers/jobseekers/job_applications_controller.rb:12]: First name: Connor, second name: Placeholder

Comment (cpjmcquillan) [reply app/controllers/jobseekers/job_applications_controller.rb:12]: I know there is a ticket coming about this so I wasn't too concerned about the content (or even where it should redirect) and that's why I didn't add a translation either. If you prefer I can just omit this bit of code completely

File: app/views/jobseekers/job_applications/build/ask_for_support.html.slim [hunk 3]
```diff
@@ -22,8 +22,8 @@
           = f.govuk_text_area :support_details, rows: 10
         = f.govuk_radio_button :support_needed, :no
 
-      = f.govuk_submit t("buttons.continue") do
-        = f.govuk_submit t("buttons.save_as_draft"), secondary: true
+      = f.govuk_submit t("buttons.continue")
+      = f.govuk_submit t("buttons.save_as_draft"), secondary: true, classes: "button-link"
```
Comment (david-mears-dfe) [inline app/views/jobseekers/job_applications/build/ask_for_support.html.slim:26]: 👍

File: app/models/job_application.rb (hunk 2 above)
Comment (cpjmcquillan) [reply app/models/job_application.rb:4]: The steps don't exist yet

File: spec/form_models/jobseekers/job_application/review_form_spec.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  context "when all steps are complete" do
+    let(:completed_steps) { JobApplication.completed_steps.keys }
+
+    it "is valid" do
+      expect(subject).to be_valid
+    end
+  end
+
+  context "when there are incomplete steps" do
+    let(:completed_steps) { JobApplication.completed_steps.keys.sample(2) }
```
Comment (david-mears-dfe) [inline spec/form_models/jobseekers/job_application/review_form_spec.rb:22]: A little voice in my head is saying "tests should be deterministic" ...

**Comment by martin-bangoura** [comment]
Review app deployed to <https://teaching-vacancies-review-pr-2959.london.cloudapps.digital>

File: spec/form_models/jobseekers/job_application/review_form_spec.rb (hunk 4 above)
Comment (cpjmcquillan) [reply spec/form_models/jobseekers/job_application/review_form_spec.rb:22]: Good point

File: app/form_models/jobseekers/job_application/review_form.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  include ActiveModel::Model
+
+  attr_accessor :confirm_data_accurate, :confirm_data_usage, :completed_steps
+
+  validates :confirm_data_accurate, acceptance: true
+  validates :confirm_data_usage, acceptance: true
+  validate :all_steps_completed
+
+  def all_steps_completed
+    return if JobApplication.completed_steps.keys.map { |step| completed_steps.include?(step) }.all?
```
Comment (david-mears-dfe) [inline app/form_models/jobseekers/job_application/review_form.rb:11]: Is this equivalent to the terser
```ruby
JobApplication.completed_steps.keys.all? { |step| completed_steps.include?(step) }
```
?

File: app/models/job_application.rb [hunk 6]
```diff
@@ -1,4 +1,16 @@
...
+    personal_details: 0,
+    professional_status: 2,
+    employment_history: 3,
+    personal_statement: 4,
+    references: 5,
+    ask_for_support: 7,
+    declarations: 8,
```
Comment (david-mears-dfe) [inline app/models/job_application.rb:11]: Philosophical question: is the 'review' step a 'step'?

File: app/form_models/jobseekers/job_application/review_form.rb (hunk 5 above)
Comment (cpjmcquillan) [reply app/form_models/jobseekers/job_application/review_form.rb:11]: Nice, didn't know `all?` could have a block!

File: app/controllers/jobseekers/job_applications_controller.rb (hunk 1 above)
Comment (david-mears-dfe) [reply app/controllers/jobseekers/job_applications_controller.rb:12]: Can you set it to `notice: "You have already applied for this job TODO: confirm content is accurate"`

**Review by david-mears-dfe — APPROVED** [review]

File: app/controllers/jobseekers/job_applications_controller.rb (hunk 1 above)
Comment (cpjmcquillan) [reply app/controllers/jobseekers/job_applications_controller.rb:12]: Actually - in hindsight I'm going to leave this part completely for https://dfedigital.atlassian.net/browse/TEVA-2036

**Pull request merged**